src/ATen/native/xpu/sycl/HistogramddKernels.cpp

```diff
@@ -24,71 +27,91 @@ struct HistogramddKernelFunctor {
   void operator()(sycl::nd_item<1> item_id) const {
     auto input_data = input_;
     int64_t wi_id = item_id.get_global_id();
-    if (wi_id < input_size_ * bin_size_) {
-      int64_t ele_idx = wi_id / bin_size_;
-      int64_t bin_idx = wi_id % bin_size_;
-
-      scalar_t value = (scalar_t)1;
-      if (use_weight_) {
-        auto weight_data = weight_;
-        value = weight_data[ele_idx];
-      }
-
-      // [left, right)
-      if (input_data[ele_idx][0] >= bin_edges_[bin_idx] &&
-          input_data[ele_idx][0] < bin_edges_[bin_idx + 1]) {
-        atomicAdd((sycl_global_ptr<scalar_t>)(hist_ + bin_idx), value);
+    if (wi_id >= input_size_) {
+      return;
+    }
+    int64_t ele_idx = wi_id;
+    int64_t hist_idx = 0;
+    for (int dim = 0; dim < input_dim_; ++dim) {
+      auto elem = input_data[ele_idx][dim];
+      const scalar_t* bin_edges = bin_edges_list_[dim];
+      auto bin_edges_size = num_bin_edges_[dim];
+      if (!(elem >= bin_edges[0] && elem <= bin_edges[bin_edges_size - 1])) {
         return;
       }
-
-      // For last bin, [left, right]
-      if (bin_idx == 0 && input_data[ele_idx][0] == bin_edges_[bin_size_]) {
-        atomicAdd((sycl_global_ptr<scalar_t>)(hist_ + bin_size_ - 1), value);
+      auto bin_idx =
+          std::upper_bound(bin_edges, bin_edges + bin_edges_size, elem) -
```

Comment:
It take much overhead to let each work item do binary search (upper_bound), and low GPU utilization. Can we consider a parallel implementation ?